spyrmsd/parallel.py

```diff
@@ -0,0 +1,118 @@
+    molrefs: Union[molecule.Molecule, List[molecule.Molecule]],
+    mols: Union[molecule.Molecule, List[molecule.Molecule]],
+    num_workers: Union[int, None] = 1,
+    symmetry: bool = True,
+    center: bool = False,
+    minimize: bool = False,
+    strip: bool = True,
+    cache: bool = True,
+    timeout: Optional[float] = None,
+    chunksize: int = 1,
+) -> List[float]:
+    """
+    Compute RMSD between two molecules with a timeout.
+    Parameters
+    ----------
+    molrefs: Union[molecule.Molecule, List[molecule.Molecule]]
+        Reference molecule
+    mols: Union[molecule.Molecule, List[molecule.Molecule]]
+        Molecules to compare to reference molecule
+    num_workers: int
+        Amount of processor to use for the parallel calculations
+    symmetry: bool, optional
+        Symmetry-corrected RMSD (using graph isomorphism)
+    center: bool, optional
+        Center molecules at origin
+    minimize: bool, optional
+        Minimised RMSD (using the quaternion polynomial method)
+    strip: bool, optional
+        Strip hydrogen atoms
+    timeout: float, optional
+        After how many seconds to stop the RMSD calculations
+    chunksize: int, optional
+        How many molecules to handle per process
+    Returns
+    -------
+    List[float]
+        RMSDs
+    """
+
+    # Ensure the num_workers is less or equal than the max number of CPUs.
+    # Silencing MyPy since os.cpu_count() can return None
+    num_workers = min(num_workers, os.cpu_count()) if os.cpu_count() is not None else 1  # type: ignore[type-var]
+
+    # Cast the molecules to lists if they aren't already
+    if not isinstance(molrefs, list):
+        molrefs = [molrefs]
+    if not isinstance(mols, list):
+        mols = [mols]
+
+    # Match the length of the molref
+    if len(molrefs) == 1 and len(molrefs) < len(mols):
+        molrefs = molrefs * len(mols)
+
+    # Ensure molrefs and mols have the same len
+    if not len(molrefs) == len(mols):
+        raise ValueError("The input mol lists have different lengths")
+
+    outputList = []
+
+    timeoutCounter = 0
+    errorCounter = 0
+
+    with ProcessPool(max_workers=num_workers) as pool:
+        rsmd_partial = partial(
+            rmsdwrapper,
+            symmetry=symmetry,
+            center=center,
+            minimize=minimize,
+            strip=strip,
+            cache=cache,
+        )
+
+        future = pool.map(
+            rsmd_partial, molrefs, mols, timeout=timeout, chunksize=chunksize
+        )
+        iterator = future.result()
+        idx = 0
+        while True:
+            try:
+                result = next(iterator)
+                outputList.append(result[0])
+            except StopIteration:
+                break
+            except TimeoutError:
+                # Not sure what the best way is to report the errors when the chunksize is non-zero
+                timeoutCounter += 1  # * chunksize
+                outputList.append(np.nan)
```

Comment:
Doesn't it need to add `chunksize` elements?

Reply:
Yes, initially it was like this, but after some testing it seems to behave slightly differently. Maybe the easiest solution would be to keep track of the number of times a chunk fails and at the end count the number of np.nan in the output and report that (along with the number of times an error was raised or a timeout occurred)?

Reply:
I don't think we need much reporting to be honest. I think just a single warning that there is at least one `np.nan` is enough, the user can then count the number of `np.nan`s if needed. What is really important is to ensure that if a whole chunk is not computed, then there are `np.nan`s for all the elements of the chunk.

Reply:
Fair enough. I agree the most important part is getting in the correct amount `np.nan`s, but this is also the trickiest, since I'm not sure if it's possible to figure out how many still were processed succesfully until the timeout hit... This "reporting" has been useful for some debugging, but I can take it out easily!

Reply:
Ah, I did not fully grasp this difficulty. So when a timeout is hit, some elements of the chunk are still computed? A safe workaround would be to only allow a chunk size of 1 with timeout, what do you think?

Reply:
> Ah, I did not fully grasp this difficulty. So when a timeout is hit, some elements of the chunk are still computed? 

From my testing yes. I put a question about this on the Pebble repo, but haven't heard back yet.
> A safe workaround would be to only allow a chunk size of 1 with timeout, what do you think?

Yeah this would make everything a lot easier!

Reply:
I completely agree! In an ideal world this wouldn't be a problem, but I think it's better to guarantee stability than risk causing unexpected bugs/errors.